File: token-client-core/src/test/java/no/nav/security/token/support/client/core/ClientPropertiesTest.java
package no.nav.security.token.support.client.core;

import com.nimbusds.oauth2.sdk.auth.ClientAuthenticationMethod;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.net.URI;
import java.util.List;

import static no.nav.security.token.support.client.core.TestUtils.jsonResponse;
import static no.nav.security.token.support.client.core.TestUtils.withMockServer;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
import static org.junit.jupiter.api.Assertions.assertNotNull;

class ClientPropertiesTest {

    private final String wellKnownJson =
        """
          {
            "issuer" : "https://someissuer",
            "token_endpoint" : "https://someissuer/token",
            "jwks_uri" : "https://someissuer/jwks",
            "grant_types_supported" : [ "urn:ietf:params:oauth:grant-type:token-exchange" ],
            "token_endpoint_auth_methods_supported" : [ "private_key_jwt" ],
            "token_endpoint_auth_signing_alg_values_supported" : [ "RS256" ],
            "subject_types_supported" : [ "public" ]
          }
          """;


    private static ClientProperties clientPropertiesFromWellKnown(URI wellKnownUrl) {
        return new ClientProperties(
            null,
            wellKnownUrl,
            OAuth2GrantType.CLIENT_CREDENTIALS,
            List.of("scope1", "scope2"),
            clientAuth(),
            null,
            tokenExchange()
        );
    }

    private static ClientAuthenticationProperties clientAuth() {
        return new ClientAuthenticationProperties(
            "client",
            ClientAuthenticationMethod.CLIENT_SECRET_BASIC,
            "secret",
            null);
    }

    private static ClientProperties.TokenExchangeProperties tokenExchange() {
        return new ClientProperties.TokenExchangeProperties(
            "aud1",
            null
        );
    }

    private static ClientProperties clientPropertiesFromGrantType(OAuth2GrantType grantType) {
        return new ClientProperties(
            URI.create("http://token"),
            null,
            grantType,
            List.of("scope1", "scope2"),
            clientAuth(),
            null,
            tokenExchange()
        );
    }

    @Test
    void validGrantTypes() {
        assertNotNull(clientPropertiesFromGrantType(OAuth2GrantType.JWT_BEARER));
        assertNotNull(clientPropertiesFromGrantType(OAuth2GrantType.CLIENT_CREDENTIALS));
        assertNotNull(clientPropertiesFromGrantType(OAuth2GrantType.TOKEN_EXCHANGE));
    }

    @Test
    void invalidGrantTypes() {
        assertThatExceptionOfType(IllegalArgumentException.class)
            .isThrownBy(() -> clientPropertiesFromGrantType(new OAuth2GrantType("somegrantNotSupported")));
    }

    @Test
    void ifWellKnownUrlIsNotNullShouldRetrieveMetadataAndSetTokenEndpoint() throws IOException {
        withMockServer(
            s -> {
                s.enqueue(jsonResponse(wellKnownJson));
                assertNotNull(clientPropertiesFromWellKnown(s.url("/well-known").uri()).getTokenEndpointUrl());
            }
        );
    }

    @Test
    void incorrectWellKnownUrlShouldThrowException(){
        assertThatExceptionOfType(OAuth2ClientException.class)
            .isThrownBy(() ->
                clientPropertiesFromWellKnown(URI.create("http://localhost:1234/notfound"))
            );
    }
}

File: token-client-core/src/test/kotlin/no/nav/security/token/support/client/core/ClientPropertiesTest.kt
package no.nav.security.token.support.client.core

import com.nimbusds.oauth2.sdk.GrantType
import com.nimbusds.oauth2.sdk.GrantType.CLIENT_CREDENTIALS
import com.nimbusds.oauth2.sdk.GrantType.JWT_BEARER
import com.nimbusds.oauth2.sdk.GrantType.TOKEN_EXCHANGE
import com.nimbusds.oauth2.sdk.auth.ClientAuthenticationMethod.CLIENT_SECRET_BASIC
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.nulls.shouldNotBeNull
import no.nav.security.token.support.client.core.ClientProperties.ClientPropertiesBuilder
import no.nav.security.token.support.client.core.ClientProperties.TokenExchangeProperties
import no.nav.security.token.support.client.core.TestUtils.jsonResponse
import no.nav.security.token.support.client.core.TestUtils.withMockServer

internal class ClientPropertiesTest : BehaviorSpec({

    Given("valid grant types") {
        When("creating client properties") {
            Then("should create valid properties for JWT_BEARER") {
                clientPropertiesFromGrantType(JWT_BEARER).shouldNotBeNull()
            }
            Then("should create valid properties for CLIENT_CREDENTIALS") {
                clientPropertiesFromGrantType(CLIENT_CREDENTIALS).shouldNotBeNull()
            }
            Then("should create valid properties for TOKEN_EXCHANGE") {
                clientPropertiesFromGrantType(TOKEN_EXCHANGE).shouldNotBeNull()
            }
        }
    }

    Given("a well-known URL is provided") {
        When("retrieving metadata") {
            Then("should set the token endpoint URL") {
                withMockServer {
                    it.enqueue(jsonResponse(wellKnownJson))
                    clientPropertiesFromWellKnown(it.url("/well-known").toString()).tokenEndpointUrl.shouldNotBeNull()
                }
            }
        }
    }

    Given("an incorrect well-known URL") {
        When("retrieving metadata") {
            Then("should throw OAuth2ClientException") {
                shouldThrow<OAuth2ClientException> {
                    clientPropertiesFromWellKnown("http://localhost:1234/notfound")
                }
            }
        }
    }

}) {
    companion object {
        private val wellKnownJson = """
          {
            "issuer" : "https://someissuer",
            "token_endpoint" : "https://someissuer/token",
            "jwks_uri" : "https://someissuer/jwks",
            "grant_types_supported" : [ "urn:ietf:params:oauth:grant-type:token-exchange" ],
            "token_endpoint_auth_methods_supported" : [ "private_key_jwt" ],
            "token_endpoint_auth_signing_alg_values_supported" : [ "RS256" ],
            "subject_types_supported" : [ "public" ]
          }
          """.trimIndent()

        private fun clientAuth() =
            ClientAuthenticationPropertiesBuilder("client", CLIENT_SECRET_BASIC)
                .clientSecret("secret")
                .build()

        private fun tokenExchange() = TokenExchangeProperties("aud1")

        fun clientPropertiesFromWellKnown(wellKnownUrl: String) =
            ClientPropertiesBuilder(CLIENT_CREDENTIALS, clientAuth())
                .wellKnownUrl(wellKnownUrl)
                .scopes("scope1", "scope2")
                .tokenExchange(tokenExchange())
                .build()

        fun clientPropertiesFromGrantType(grantType: GrantType) =
            ClientPropertiesBuilder(grantType, clientAuth())
                .tokenEndpointUrl("http://token")
                .scopes("scope1", "scope2")
                .tokenExchange(tokenExchange())
                .build()
    }
}
